### PR TITLE
feat(clustering): hybrid forward proxy https

### PR DESCRIPTION
**Pull request by fffonion**
Add support of talk to an HTTP tunnel in https

FTI-2996

### CODE REVIEW

File: kong/resty/websocket/client.lua [hunk 1]
```diff
@@ -159,22 +160,27 @@ function _M.connect(self, uri, opts)
...
+                return nil, "error parsing proxy_url: " .. err
+
+            elseif m[1] ~= "http" and m[1] ~= "https" then
+                return nil, "only proxy with scheme \"http\" or \"https\" is supported"
+            end
 
-        elseif m[1] ~= "http" then
-            return nil, "only HTTP proxy is supported"
+            connect_host = m[2]
+            connect_port = m[3] or 80 -- hardcode for now as we only support HTTP proxy
```
Comment (dndx) [inline kong/resty/websocket/client.lua]: We are supporting TLS now, so this probably doesn't makes sense anymore.

File: kong/clustering/utils.lua [hunk 2]
```diff
@@ -335,10 +339,16 @@ local function parse_proxy_url(conf)
...
     -- assume proxy_server is validated in conf_loader
     local parsed = parse_url(proxy_server)
-    ret.proxy_url = fmt("%s://%s:%s", parsed.scheme, parsed.host, parsed.port or 443)
-    ret.scheme = parsed.scheme
-    ret.host = parsed.host
-    ret.port = parsed.port
+    if parsed.scheme == "https" then
+      ret.proxy_url = CLUSTER_PROXY_SSL_TERMINATOR_SOCK
+      -- hide other fields to avoid it being accidently used
+      -- the connection details is statically rendered in nginx template
```
Comment (dndx) [inline kong/clustering/utils.lua:345]:
```suggestion
      -- the connection details is statically rendered in nginx template

```

File: kong/resty/websocket/client.lua [hunk 3]
```diff
@@ -159,22 +160,27 @@ function _M.connect(self, uri, opts)
...
-        local m, err = re_match(
-            proxy_url,
-            [[^(?:(http[s]?):)?//((?:[^\[\]:/\?]+)|(?:\[.+\]))(?::(\d+))?([^\?]*)\??(.*)]],
-            "jo"
-        )
-        if err then
-            return nil, "error parsing proxy_url: " .. err
+        if str_sub(proxy_url, 1, 6) == "unix:/" then
+            connect_host = proxy_url
+            connect_port = nil
```
Comment (dndx) [inline kong/resty/websocket/client.lua:165]:
```suggestion
            connect_port = nil

```

File: spec/02-integration/09-hybrid_mode/10-forward-proxy_spec.lua [hunk 4]
```diff
@@ -8,7 +8,14 @@ local fixtures = {
...
     server {
       listen 16797;
+      listen 16799 ssl;
+      listen [::]:16799 ssl;
+
+      ssl_certificate ../spec/fixtures/kong_spec.crt;
+      ssl_certificate_key ../spec/fixtures/kong_spec.key;
+
       error_log logs/proxy.log debug;
+      error_log logs/proxy2.log debug;
```
Comment (fffonion) [inline spec/02-integration/09-hybrid_mode/10-forward-proxy_spec.lua:18]:
```suggestion
```

File: spec/02-integration/09-hybrid_mode/10-forward-proxy_spec.lua [hunk 5]
```diff
@@ -17,29 +24,58 @@ local fixtures = {
...
     server {
       listen 16796;
+      listen 16798 ssl;
+      listen [::]:16798 ssl;
+
+      ssl_certificate ../spec/fixtures/kong_spec.crt;
+      ssl_certificate_key ../spec/fixtures/kong_spec.key;
+
       error_log logs/proxy_auth.log debug;
+      error_log logs/proxy_auth2.log debug;
```
Comment (fffonion) [inline spec/02-integration/09-hybrid_mode/10-forward-proxy_spec.lua:34]:
```suggestion
```

File: kong/conf_loader/init.lua [hunk 6]
```diff
@@ -1861,6 +1862,15 @@ local function load(path, custom_conf, opts)
 
   log.verbose("prefix in use: %s", conf.prefix)
 
+  -- hybrid mode HTTP tunneling (CONNECT) proxy inside HTTPS
+  if conf.cluster_use_proxy then
+    -- throw err, assume it's already handled in check_and_infer
+    local parsed = assert(require("socket.url").parse(conf.proxy_server))
```
Comment (chronolaw) [inline kong/conf_loader/init.lua:1868]: localize this `require("socket.url").parse`?

**Review by dndx — APPROVED** [review]
This looks fine, can be merged after all comments are addressed.

**Pull request merged**